storage-api/src/test/org/apache/hadoop/hive/ql/exec/vector/TestBytesColumnVector.java

```diff
@@ -74,16 +88,97 @@ public void testSmallBufferReuse() {
 
     // All small writes to the first buffer should be in contiguous memory
     for (int i = 0; i < bytesWrittenToBytes1; ++i) {
-      assertEquals((byte) 1, bytes1[i]);
+      assertEquals((byte) (i / smallWriteSize + 1), bytes1[i]);
+    }
+  }
+
+  /**
+   * Test the setVal, setConcat, and StringExpr.padRight methods.
+   */
+  @Test
+  public void testConcatAndPadding() {
+    BytesColumnVector col = new BytesColumnVector();
+    col.reset();
+    byte[] prefix = "緑".getBytes(StandardCharsets.UTF_8);
+
+    // fill the column with 'test'
+    for(int row=0; row < col.vector.length; ++row) {
+      col.setVal(row, prefix, 0, prefix.length);
+    }
+    for(int row=0; row < col.vector.length; ++row) {
+      assertEquals("row " + row, "緑", col.toString(row));
+    }
+
+    // pad out to 6 characters
+    for(int row=0; row < col.vector.length; ++row) {
+      StringExpr.padRight(col, row, col.vector[row], col.start[row],
+          col.length[row], 6);
+    }
+    for(int row=0; row < col.vector.length; ++row) {
+      assertEquals("row " + row, "緑     ", col.toString(row));
+    }
+
+    // concat the row digits
+    for(int row=0; row < col.vector.length; ++row) {
+      byte[] rowStr = Integer.toString(row).getBytes(StandardCharsets.UTF_8);
+      col.setConcat(row, col.vector[row], col.start[row], col.length[row],
+          rowStr, 0, rowStr.length);
+    }
+    for(int row=0; row < col.vector.length; ++row) {
+      assertEquals("row " + row, "緑     " + row, col.toString(row));
+    }
+
+    // We end up allocating 20k, so we should have expanded the small buffer
+    assertEquals(32 * 1024, col.bufferSize());
+  }
+
+  @Test
+  public void testBufferOverflow() {
+    BytesColumnVector col = new BytesColumnVector(2048);
+    col.reset();
+    assertEquals(BytesColumnVector.DEFAULT_BUFFER_SIZE, col.bufferSize());
+
+    // pick a size below 1m so that we use the small buffer;
+    final int size = BytesColumnVector.MAX_SIZE_FOR_SMALL_ITEM - 1024;
+
+    // run through once to expand the small value buffer
+    for(int row=0; row < col.vector.length; ++row) {
+      writeToBytesColumnVector(row, col, size, row);
+    }
+    // it should have resized a bunch of times
+    byte[] smallBuffer = col.getValPreallocatedBytes();
+    assertNotSame(smallBuffer, col.vector[0]);
+    assertSame(smallBuffer, col.vector[1024]);
+
+    // reset the column, but make sure the buffer isn't reallocated
+    col.reset();
+    assertEquals(BytesColumnVector.MAX_SIZE_FOR_SMALL_BUFFER, col.bufferSize());
+
+    // fill up the vector now with the large buffer
+    for(int row=0; row < col.vector.length; ++row) {
+      writeToBytesColumnVector(row, col, size, row);
+    }
+    assertEquals(BytesColumnVector.MAX_SIZE_FOR_SMALL_BUFFER, col.bufferSize());
+    // now the first 1025 rows should all be the small buffer
```

Comment:
I would explicitly mention that the reason is: 1025 *  MAX_SIZE_FOR_SMALL_ITEM is below MAX_SIZE_FOR_SMALL_BUFFER